### PR TITLE
Partitioner Feature Additions: Mapping generation, net-span partition report, LUTs/partition report, verbose hypergraph reports, LUTCount CLI, mapping constraint partition input

### PR DESCRIPTION
**Pull request by pnewlin-amd**
A series of QoL features and additions to the partitioner branch.

- Added hierarchical mapping output (mapping.txt), if an entire submodule `TOP/Counter` is wholly contained in partition 1 (FPGA_B) then produce only one line `TOP/Counter FPGA_B`
- Added _optional_ human readable hypergraph reports. After a partition a user can easily see which nets have been cut and their names.
- Added a dedicated LUT report .txt file output which shows how many LUTs have been assigned per partition.
- Exposed KaHyPar hypergraph switches to command line interface
- Added partition constraints, this allows the user to specify a mapping constraint which KaHyPar will interpret as a .fix input for assigning some submodule to exactly one partition.  
- Changed partition naming to be FPGA_<...>.
- Added a cells.txt output which displays every partition name. 
- Added command line interface option to specify output location of partition.
- Relocated the LUT counting functionality to a single location (source of truth)
- Added verbose logging to partition process to display caught edge cases and memory usage.
- Enabled KaHyPar verbose timing output.
- Added partition support for .DCP input.
- Added read netlist command line interface tool. Allows user to test memory usage on monolithic netlists.
- Added LUT count command line interface tool. Allows user to provide a netlist and RapidWright will return total LUT count of top level.

### CODE REVIEW

File: src/com/xilinx/rapidwright/debug/memory/audit/ReadNetlist.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.io.File;
+
+import com.xilinx.rapidwright.design.Design;
+import com.xilinx.rapidwright.edif.EDIFTools;
+
+/**
+ * Memory audit utility for reading netlists.
+ */
+public class ReadNetlist {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/debug/memory/audit/ReadNetlist.java:33]: This class seems to duplicate functionality in the [`CodePerfTracker`](https://github.com/Xilinx/RapidWright/blob/master/src/com/xilinx/rapidwright/tests/CodePerfTracker.java) class.  This class is employed when loading a DCP and, when set with the right options, can produce a similar report.  When calling `Design.readCheckpoint()` there are overload options to provide a custom configured `CodePerfTracker` instance.  Have a look to see if that would meet your needs.

I'm thinking that having something like a `DesignAnalyzer` that can provide an array of analysis options while simply loading a design would be useful, but I'm not sure this is the right PR for it.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+    /**
+     * Trie node representing a hierarchical instance path segment.
+     *
+     * - name: the segment at this level (for example, "stage_16").
+     * - children: sub-segments under this node.
+     * - leafCounts: one count per partition per explicit path)
+     * - counts: aggregated counts over this node’s subtree (leafCounts + all children)
+     *           lets us easily decide if wholly contained or not
+     */
+    private static final class Node {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:68]: `Node` is an overloaded term.  Perhaps we could refactor to something more unique such as `HierNode`, or something similar.

File: src/com/xilinx/rapidwright/util/LUTCount.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import com.xilinx.rapidwright.edif.EDIFCell;
+import com.xilinx.rapidwright.edif.EDIFHierCellInst;
+import com.xilinx.rapidwright.edif.EDIFNetlist;
+import com.xilinx.rapidwright.edif.EDIFTools;
+import com.xilinx.rapidwright.edif.partition.PartitionTools;
+
+/**
+ * CLI utility to count logic LUT usage from EDIF (.edf/.edif) or Vivado DCP (.dcp).
+ */
+public final class LUTCount {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/util/LUTCount.java:40]: I think the functionality of this class should be absorbed into a `DesignAnalyzer` class.  See my comment on `ReadNetlist.java`.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+     */
+    public static void writeMappingFile(List<String> lines, String outputPath) {
+        try {
+            Path out = Paths.get(outputPath);
+            Path parent = out.getParent();
+            if (parent != null) {
+                Files.createDirectories(parent);
+            }
+            Files.write(out, lines, StandardCharsets.UTF_8);
+        } catch (IOException e) { // more guards the better.
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:167]: Not sure what you mean by this comment?

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 5]
```diff
@@ -0,0 +1,746 @@
...
+
+    /**
+     * RapidWright partitioner api: writes mapping.txt in outDir.
+     *
+     * @param outDir The output directory.
+     * @param netlist The EDIF netlist.
+     * @param partitions A map of partition IDs to sets of instance names.
+     * @param instLutCountMap A map of instance LUT counts (unused but kept for API consistency).
+     */
+    public static void write(Path outDir
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:180]: This method is confusingly named as the Javadoc says it writes the mapping file, but there is a method just above it that is called `writeMappingFile()`.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 6]
```diff
@@ -0,0 +1,746 @@
...
+        return partToPaths;
+    }
+
+    /**
+     * Safe read of file lines with utf-8.
+     *
+     * @param file The file to read.
+     * @return A list of lines from the file.
+     */
+    private static List<String> safeReadLines(Path file) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:462]: There is a similar method in `FileTools`: https://github.com/Xilinx/RapidWright/blob/57438b963079ad74ed29d476a25e7486c5222776/src/com/xilinx/rapidwright/util/FileTools.java#L663

Perhaps we should add an overload there to take a `Path`.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 7]
```diff
@@ -0,0 +1,746 @@
...
+
+    /**
+     * Case-insensitive search for a file that matches either "<name>" or "<name>.txt"
+     * in a directory.
+     *
+     * @param dir The directory to search in.
+     * @param targetName The name to search for.
+     * @return The path to the matching file, or null if not found.
+     */
+    private static Path findFileCaseInsensitive(Path dir, String targetName) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:478]: Is this method really necessary?

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java (hunk 2 above)
Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java]: Also, let's move it into its own file, there are several methods at the end of this file that could be refactored to live in that class.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 8]
```diff
@@ -0,0 +1,746 @@
...
+        return null;
+    }
+
+    /**
+     * Trims comments and whitespace; treats lines starting with '#' or '//' as comments.
+     *
+     * @param s The string to trim.
+     * @return The trimmed string, or an empty string if it was a comment.
+     */
+    private static String trimComment(String s) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:501]: This feels like maybe it should be in a more generic file, or perhaps could be replaced by stream filters.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+    }
+
+    /**
+     * Returns lines sorted by path depth, then lexicographically. Useful when post-processing
+     * is needed for readability.
+     *
+     * @param lines The list of lines to sort.
+     * @return A new sorted list of lines.
+     */
+    public static List<String> sortByPathDepth(List<String> lines) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:773]: This is not used outside the class, should we make it `private`?

File: src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+package com.xilinx.rapidwright.edif.partition;
+
+import com.xilinx.rapidwright.edif.EDIFHierCellInst;
+
+/**
+ * logicDiscoveryPolicy is a source-of-truth utility for
+ * discovering logic-related properties in the EDIF/DCP domain.
+ * For example, determining if a cell is a look up table.
+ */
+public final class logicDiscoveryPolicy {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java:33]: Class names should begin with a capital letter.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+        // append a single top 'home' line at the end for readability.
+        // ============================================================================
+        String topName = null;
+        try {
+            EDIFHierCellInst topHierCellInst = netlist.getTopHierCellInst();
+            topName = (topHierCellInst != null) ? topHierCellInst.getCellName() : null;
+            if (topName == null || topName.isEmpty()) {
+                topName = (topHierCellInst != null) ? topHierCellInst.getInst().getName() : null;
+            }
+        } catch (Exception ignored) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:256]: Having a blanket catch for all exceptions is generally frowned upon.  What is happening here that requires such usage?

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 12]
```diff
@@ -0,0 +1,746 @@
...
+        if (prefixed.size() > 10) {
+            PartitionTools.logPartitionerDebug(null,
+                    "MAPPING_ORDER_DEBUG: post-sort ... (middle lines omitted)");
+            for (int i = prefixed.size() - showLimit; i < prefixed.size(); i++) {
+                PartitionTools.logPartitionerDebug(null,
+                        "MAPPING_ORDER_DEBUG: post-sort[%d] depth=%d line=%s",
+                        i, depthOfPath(prefixed.get(i)), prefixed.get(i));
+            }
+        }
+        java.util.Map<Integer, Integer> depthCounts = new java.util.HashMap<>();
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:317]: This should just be imported at the top the file rather than in context.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 13]
```diff
@@ -0,0 +1,746 @@
...
+                Path matched = findFileCaseInsensitive(partitionDir, p);
+                if (matched != null && Files.isRegularFile(matched)) {
+                    readOk = true;
+                    for (String line : safeReadLines(matched)) {
+                        String trimmedLine = trimComment(line);
+                        if (trimmedLine.isEmpty()) {
+                            continue;
+                        }
+                        paths.add(trimmedLine);
+                    }
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:440]: There is a common pattern occurring here 3 times in a row.  Perhaps we could collapse it down into a single method (or modify the existing methods that are called).  It would be great if we could make it more generic so that it could be used elsewhere.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+        cur.leafCounts.put(partition, cur.leafCounts.getOrDefault(partition, 0) + 1);
+    }
+
+    /**
+     * Splits a hierarchical path on '/' while leaving each segment intact.
+     *
+     * @param path The path to split.
+     * @return A list of path segments.
+     */
+    private static List<String> splitPath(String path) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:566]: Not sure that this deserves to be its own method.

Comment (clavin-xlnx) [reply src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:566]: Also, it turns out that it is valid to have `/` as part of a name of a hierarchical cell.  Not sure if your designs will encounter that, we regularly we find this to be the case from IP from Vivado.  We do have methods that will compensate for that (see `EDIFNetlist.getHierObject()`).

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+    }
+
+    /**
+     * Computes per-subtree aggregated counts for every node in the trie.
+     *
+     * @param node The root node of the subtree.
+     * @return A map of aggregated counts per partition.
+     */
+    private static Map<String, Integer> computeCounts(Node node) {
+        Map<String, Integer> aggregatedCounts = new LinkedHashMap<>();
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:581]: I've noticed a lot of `LinkedHashMap` usage, do you need ordered traversal in all those cases?

File: src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+    // List of protected names that should be excluded from mapping generation. 
+    // GND and VCC will be re-created locally in the partition / module as needed,
+    // GND, VCC should not be results you see in mapping.txt
+    private static final java.util.LinkedHashSet<String> EXCLUSION_NAMES =
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java:38]: The class should be imported at the top instead of using fully qualified name in context.

File: src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+     * Returns true if a cell type name represents a LUT used as logic.
+     * 
+     * Treat any cell type whose name contains "LUT" as a logic LUT.
+     * substring match to be consistent with legacy policy.
+     * 
+     * TODO : More concrete & fact-checked discovery policy?
+     * @param type_name The cell type name to check
+     * @return True if the type name contains "LUT"
+     */
+    public static boolean isLogicLUT(String type_name) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/logicDiscoveryPolicy.java:58]: Can you use `LUTTools.isCellALUT()` instead?

File: src/com/xilinx/rapidwright/edif/partition/LUTCountingUtils.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Recursively computes LUT count for hierarchical cell instance. Utilize caching 
+     * to avoid recalculating known cell types.
+     * 
+     * @param hierInst Hierarchical cell instance
+     * @param cellTypeCache Cell type cache
+     * @param instMap Instance map
+     * @return LUT count
+     */
+    public static Integer computeLUTCount(EDIFHierCellInst hierInst,
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/LUTCountingUtils.java:84]: We should probably add a test for this method, let's connect later about the best way to create one.

File: src/com/xilinx/rapidwright/edif/partition/MtKaHyParPartitioner.java [hunk 19]
```diff
@@ -42,67 +42,111 @@ public class MtKaHyParPartitioner implements AbstractPartitioner {
...
+    /**
+     * Executes the MtKaHyPar partitioning tool with configured parameters.
+     * @return Exit code from the partitioner execution
+     */
     @Override
     public Integer runPartitioner() {
+        // If fixed vertices are present, override incompatible presets
+        String effectivePreset = presetType;
+        if (fixedVerticesFile != null) {
+            if ("deterministic".equals(effectivePreset) || "large_k".equals(effectivePreset)) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/MtKaHyParPartitioner.java:66]: We should probably create class member `public static final String`s for the two static strings here.

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java [hunk 20]
```diff
@@ -47,89 +51,777 @@
  */
 public class Partitioner {
 
+    /**
+     * Logs current memory usage for debugging.
+     * @param label Description label for the audit point
+     */
+    private static void memAudit(String label) {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/Partitioner.java:58]: Again, let's try to use `CodePerfTracker` to avoid code duplication.

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java [hunk 21]
```diff
@@ -47,89 +51,777 @@
...
+     */
     public static AbstractPartitioner getDefaultPartitioner() {
         MtKaHyParPartitioner p = new MtKaHyParPartitioner();
         return p;
     }
+
+    /**
+     * Prints command-line usage information to console.
+     */
+    private static void printUsage() {
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/Partitioner.java:81]: This has enough options that I think we should probably switch this to use the `OptionParser` class.  It will lead to a more maintainable solution and consistency over time.  You can take a look at `PerformanceExplorer` as an example.

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java [hunk 22]
```diff
@@ -47,89 +51,777 @@
...
         CodePerfTracker t = new CodePerfTracker("Partitioner");
+        
+        boolean generateEdifNets = false; // Human readable hypergraph artifacts
+        Path constraintsFile = null;
+        boolean constraintsDebug = false;
+        boolean skipDetailedReports = false;
+        
+
+
+
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/Partitioner.java:127]: Remove extra whitespace.

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java [hunk 23]
```diff
@@ -47,89 +51,777 @@
...
+        // ============================================================================
+        // ARGUMENT PARSING & INITIALIZATION
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/Partitioner.java:129]: It seems like `main()` is broken up into various tasks, but they are all in the same method (its over 700 lines long).  This should be broken up into separate methods.

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java [hunk 24]
```diff
@@ -47,89 +51,777 @@
...
+        }
+        if (inLower.endsWith(".dcp")) {
+            PartitionTools.logPartitionerDebug(
+                    System.err,
+                    "DCP backfill start -> leafCount=%d",
+                    leafInsts.size());
+            int backfillAttempts = 0;
+            int backfillSuccess = 0;
+            int backfillSkipped = 0;
+            java.util.Map<com.xilinx.rapidwright.edif.EDIFCell, java.lang.Integer> lutCountCache;
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/Partitioner.java:244]: Please avoid fully qualified references and instead just import the class at the top of the file.

File: src/com/xilinx/rapidwright/edif/partition/PartitionLabel.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/PartitionLabel.java:1]: Not sure if this functionality needs its own class.  Perhaps refactor to make it a bit more generic and put it in `StringTools`?

File: src/com/xilinx/rapidwright/edif/partition/PartitionPairCounter.java [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+        List<String> keys = new ArrayList<>(pairCounts.keySet());
+        Collections.sort(keys);
+        List<String> lines = new ArrayList<>(keys.size());
+        for (String pairKey : keys) {
+            int separatorIndex = pairKey.indexOf(',');
+            String partitionLabelA = pairKey.substring(0, separatorIndex);
+            String partitionLabelB = pairKey.substring(separatorIndex + 1);
+            int pairCount = pairCounts.get(pairKey);
+            
+            //format will be "FPGA_A--count--FPGA_B".
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/edif/partition/PartitionPairCounter.java:153]:
```suggestion
            // format will be "FPGA_A--count--FPGA_B".
```

**Review by clavin-xlnx — CHANGES_REQUESTED** [review]
First round of feedback, please reach out with questions.  I didn't make comments on every occurrence, but hopefully you'll be able to apply the patterns across all the changes.

Thanks for the PR!

File: src/com/xilinx/rapidwright/edif/partition/Partitioner.java (hunk 23 above)
Comment (pnewlin-amd) [reply src/com/xilinx/rapidwright/edif/partition/Partitioner.java:129]: Good point. Next commit will include `PartitionSolutionProcessor.java` all of the processing and artifact generation (mapping.txt, lut_report.txt, etc.) that occurs after the KaHyPar partition solution is found will be moved into this file. Partitioner.java will utilize these functions.

Going to move netlist coarsening and edge discovery code into `partitiontools.java`.

Partitioner.java is sitting at ~400lines on my side right now, hopefully this next commit makes organization much easier to follow.

File: src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java (hunk 6 above)
Comment (pnewlin-amd) [reply src/com/xilinx/rapidwright/edif/partition/HierMappingWriter.java:462]: > There is a similar method in `FileTools`:
> 
> https://github.com/Xilinx/RapidWright/blob/57438b963079ad74ed29d476a25e7486c5222776/src/com/xilinx/rapidwright/util/FileTools.java#L663
> 
> Perhaps we should add an overload there to take a `Path`.

I can include this in next commit? -->

```Java
    /**
     * This is a simple method that will read in a text file and put each line in a
     * string and put all the lines in an ArrayList.  The user is cautioned not
     * to open extremely large files with this method.
     * @param fileName Name of the text file to load.
     * @return An ArrayList containing strings of each line
     */
    public static ArrayList<String> getLinesFromTextFile(Path fileName) {
        return getLinesFromTextFile(fileName.toString());
    }
```

**Comment by pnewlin-amd** [comment]
This feedback has been incorporated into a new RapidWright branch. Goal of future PR is to only have code changes that affect partitioning / `mapping.txt` output, many of the additional artifacts this branch produces are unnecessary and add far too much bloat & code.
